src/context.js

```diff
@@ -0,0 +1,191 @@
+    let child = new ExecutionContext(this, options);
+    this.children.add(child);
+    child.ensure(() => {
+      this.children.delete(child);
+      let { result } = child;
+      if (this.isRunning || this.isWaiting) {
+        if (child.isErrored) {
+          if (typeof options.fail === 'function') {
+            options.fail(result, this.fail);
+          } else {
+            this.fail(result);
+          }
+        } else if (child.isHalted) {
+          if (typeof options.halt === 'function') {
+            options.halt(result);
+          } else {
+            this.fail(new HaltError(result));
+          }
+        }
+        if (this.isWaiting && !this.hasBlockingChildren) {
+          this.finalize('completed');
+        }
+      }
+    });
+    child.enter(operation);
+    return child;
+  }
+
+  ensure(hook) {
+    let { exitHooks } = this;
+    exitHooks.add(hook);
+    return () => exitHooks.delete(hook);
+  }
+
+  enter(operation) {
+    if (this.isUnstarted) {
+      let controller = this.createController(operation);
+      this.operation = operation;
+      this.state = 'running';
+
+      let { resume, fail, ensure, call } = this;
+      controller.call({ resume, fail, ensure, call, context: this });
+
+    } else {
+      throw new Error(`
+Tried to call #enter() on a Context that has already been finalized. This
+should never happen and so is almost assuredly a bug in effection. All of
+its users would be in your eternal debt were you to please take the time to
+report this issue here:
+https://github.com/thefrontside/effection.js/issues/new
+
+Thanks!`);
+    }
+  }
+
+  halt(reason) {
+    if (this.isBlocking) {
+      this.finalize('halted', reason);
+    }
+  }
+
+  resume(value) {
+    if (this.isRunning) {
+      this.result = value;
+      if (this.hasBlockingChildren) {
+        this.state = 'waiting';
+      } else {
+        this.finalize('completed', value);
+      }
+    } else {
+      throw new Error(`
+Tried to call Fork#resume() on a Fork with state '${this.state}' This
+should never happen and so is almost assuredly a bug in effection. All of
+its users would be in your eternal debt were you to please take the time to
+report this issue here:
+https://github.com/thefrontside/effection.js/issues/new
+
+Thanks!`);
+    }
+    if (this.options.resume) {
+      this.options.resume(value);
+    }
+  }
+
+  fail(error) {
+    this.finalize('errored', error);
+  }
+
+  finalize(state, result) {
+    this.result = result || this.result;
+    this.state = state;
+    for (let hook of this.exitHooks) {
+      this.exitHooks.delete(hook);
+      hook(this);
+    }
+    for (let child of this.children) {
+      this.children.delete(child);
+      child.halt(result);
+    }
+    this.finalizePromise();
+  }
+
+  createController(operation) {
+    let controller = ControlFunction.for(operation);
+    if (!controller) {
+      throw new Error(`cannot find controller for ${operation}`);
+    }
```

Comment:
This is how we can make operations and effects super-general. Instead of just giving up if we don't find the controller for a given operation, we can search _upwards_ on the stack of execution contexts, checking to see at each point if it can handle the operation.